### PR TITLE
Handle empty screenhosts and trailers

### PR DESCRIPTION
**Pull request by olegoid**
If bundle has no screenshots and/or trailers attached parsing fails
with "undefined method `fetch' for nil:NilClass". I added a couple
of nil checks to handle this corner case properly.

Related issue: https://github.com/fastlane/fastlane/issues/5701

### CODE REVIEW

File: spaceship/lib/spaceship/tunes/app_version.rb [hunk 1]
```diff
@@ -557,7 +557,10 @@ def setup_screenshots
 
       # generates the nested data structure to represent screenshots
       def setup_screenshots_for(row)
-        display_families = row.fetch("displayFamilies", {}).fetch("value", nil)
+        display_families_hash = row.fetch("displayFamilies", {})
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/tunes/app_version.rb]: If the first row.fetch would be `nil`, it should fallback to `{}` as the default value. I think `row` might be nil at this point, causing the crash.

Have you actually tried this fix locally? Could you do so, and if it doesn't work, revert the changes, and add a new line as the first line of the method
```ruby
def setup_screenshots_for(row)
  return if row.nil?
  display_famililes ...
```

Comment (olegoid) [inline spaceship/lib/spaceship/tunes/app_version.rb]: Yes, I tried this fix locally and it worked for me. However, your suggestion to change fallback type and check row for nil definitely make sense. Thanks!

File: spaceship/lib/spaceship/tunes/app_version.rb [hunk 2]
```diff
@@ -628,7 +633,12 @@ def setup_trailers
 
       # generates the nested data structure to represent trailers
       def setup_trailers_for(row)
-        display_families = row.fetch("displayFamilies", {}).fetch("value", nil)
+        return if row.nil?
+
+        display_families_hash = row.fetch("displayFamilies", {})
+        return {} unless display_families_hash
```
Comment (mfurtak) [inline spaceship/lib/spaceship/tunes/app_version.rb]: Hm, I don't think we want this method to possibly return `{}`.

With your change  below, I think you can adjust the line above to be:

```ruby
display_families_hash = row.fetch("displayFamilies", {}) || {}
```

and remove the line starting with `return`

File: spaceship/lib/spaceship/tunes/app_version.rb [hunk 3]
```diff
@@ -557,7 +557,12 @@ def setup_screenshots
 
       # generates the nested data structure to represent screenshots
       def setup_screenshots_for(row)
-        display_families = row.fetch("displayFamilies", {}).fetch("value", nil)
+        return if row.nil?
+
+        display_families_hash = row.fetch("displayFamilies", {})
+        return {} unless display_families_hash
```
Comment (mfurtak) [inline spaceship/lib/spaceship/tunes/app_version.rb]: Hm, I don't think we want this method to possibly return {}.

With your change below, I think you can adjust the line above to be:

```ruby
display_families_hash = row.fetch("displayFamilies", {}) || {}
```

and remove the line starting with return

**Comment by mfurtak** [comment]
As a note for us, the more I see how `Hash#fetch` behaves, the less I like it. It's easy to think that it will prevent return of a `nil` value, but if the key is present and has a `nil` value, that's what you wind up getting...

**Comment by KrauseFx** [comment]
Nice catch @mfurtak - I didn't know that this is how `.fetch` works when the value is nil.

**Comment by mfurtak** [comment]
Yeah, I just noticed too 😅

**Comment by yukw777** [comment]
Thanks for the quick fix! Sorry to bug you guys, but any estimate on when this would be merged and a new version would be released?

**Comment by mfurtak** [comment]
@yukw777 Sorry, it's probably not realistic to give an estimate, but we'll try to get it out as soon as possible! 👍

**Comment by KrauseFx** [comment]
Sorry, unfortunately I can't reproduce the error. I'll dive into the code 👍

File: spaceship/lib/spaceship/tunes/app_version.rb (hunk 2 above)
Comment (KrauseFx) [inline spaceship/lib/spaceship/tunes/app_version.rb]: This should be [] instead of {} yeah

**Comment by KrauseFx** [comment]
Hey @olegoid, since I can't reproduce, could you help me test the following PR locally https://github.com/fastlane/fastlane/pull/5741

```
git fetch origin
git checkout -b handle-empty-screenshots-and-trailers origin/handle-empty-screenshots-and-trailers
```
And then install spaceship locally using
```
bundle update
rake install
```
and try running spaceship again 👍

**Comment by adomanico** [comment]
Seeing this as well. Thanks for the quick response

**Comment by KrauseFx** [comment]
@adomanico Is there any way for me to reproduce the error? I tried with an empty app, but it works just fine. 
Could you give my patch a try? If the patch fixes the error I can prepare a new release

**Comment by adomanico** [comment]
Ran lane with bundle exec

Failed on deliver with these params:
```
deliver(username: "REDACTED",
     app_identifier: "REDACTED",
     ipa: "./build/APPNAME-#{buildNumber}.ipa",
     team_id: "REDACTED",
     app_version: "#{versionNumber}",
     force: true,
     skip_screenshots: true,
     skip_metadata: true,
     skip_binary_upload: false,
     submit_for_review: false,
     automatic_release: false)
```

**Comment by KrauseFx** [comment]
@adomanico can you share the complete stack trace and the output (you can get the stack trace when running using `--verbose`) 👍

**Comment by adomanico** [comment]
```
ERROR [2016-08-14 20:56:29.93]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...
Search query: undefined method `fetch' for nil

URL: https://api.github.com/search/issues?q=undefined%20method%20%60fetch'%20for%20nil+repo:fastlane/fastlane
➡️  undefined method `fetch' for nil:NilClass
   https://github.com/fastlane/fastlane/issues/5677 [open] 1 💬
   2 days ago

➡️  Spaceship parsing fails with "undefined method `fetch' for nil:NilClass"
   https://github.com/fastlane/fastlane/issues/5701 [open] 2 💬
   2 days ago

➡️  delivery init/run error:  `setup_screenshots_for': [!] undefined method `fetch' for nil:NilClass (NoMethodError)
   https://github.com/fastlane/fastlane/issues/5643 [closed] 26 💬
   5 days ago

and 37 more at: https://github.com/fastlane/fastlane/search?q=undefined method `fetch' for nil&type=Issues&utf8=✓
bundler: failed to load command: fastlane (/usr/local/bin/fastlane)
NoMethodError: [!] undefined method `fetch' for nil:NilClass
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/app_version.rb:546:in `setup_screenshots_for'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/app_version.rb:540:in `block in setup_screenshots'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/app_version.rb:538:in `each'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/app_version.rb:538:in `setup_screenshots'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/app_version.rb:303:in `setup'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/base.rb:182:in `initialize'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/app_version.rb:158:in `new'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/app_version.rb:158:in `factory'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/app_version.rb:180:in `find'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/application.rb:102:in `edit_version'
  /Library/Ruby/Gems/2.0.0/gems/spaceship-0.29.1/lib/spaceship/tunes/application.rb:168:in `ensure_version!'
  /Library/Ruby/Gems/2.0.0/gems/deliver-1.13.1/lib/deliver/runner.rb:39:in `verify_version'
  /Library/Ruby/Gems/2.0.0/gems/deliver-1.13.1/lib/deliver/runner.rb:20:in `run'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/actions/deliver.rb:18:in `run'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/runner.rb:192:in `block (2 levels) in execute_action'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/actions/actions_helper.rb:35:in `execute_action'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/runner.rb:177:in `block in execute_action'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/runner.rb:176:in `chdir'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/runner.rb:176:in `execute_action'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/runner.rb:102:in `trigger_action_by_name'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/fast_file.rb:140:in `method_missing'
  Fastfile:84:in `block (2 levels) in parsing_binding'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/lane.rb:33:in `call'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/lane.rb:33:in `call'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/runner.rb:50:in `block in execute'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/runner.rb:46:in `chdir'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/runner.rb:46:in `execute'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/lane_manager.rb:46:in `cruise_lane'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/command_line_handler.rb:30:in `handle'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/commands_generator.rb:51:in `block (2 levels) in run'
  /Library/Ruby/Gems/2.0.0/gems/commander-4.3.5/lib/commander/command.rb:178:in `call'
  /Library/Ruby/Gems/2.0.0/gems/commander-4.3.5/lib/commander/command.rb:178:in `call'
  /Library/Ruby/Gems/2.0.0/gems/commander-4.3.5/lib/commander/command.rb:153:in `run'
  /Library/Ruby/Gems/2.0.0/gems/commander-4.3.5/lib/commander/runner.rb:428:in `run_active_command'
  /Library/Ruby/Gems/2.0.0/gems/fastlane_core-0.50.0/lib/fastlane_core/ui/fastlane_runner.rb:36:in `run!'
  /Library/Ruby/Gems/2.0.0/gems/commander-4.3.5/lib/commander/delegates.rb:15:in `run!'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/commands_generator.rb:248:in `run'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/commands_generator.rb:20:in `start'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/lib/fastlane/cli_tools_distributor.rb:53:in `take_off'
  /Library/Ruby/Gems/2.0.0/gems/fastlane-1.100.0/bin/fastlane:5:in `<top (required)>'
  /usr/local/bin/fastlane:23:in `load'
  /usr/local/bin/fastlane:23:in `<top (required)>'
```
